src/afrolabs/components/kafka.clj

```diff
@@ -1750,76 +1892,206 @@
+                                     ;; nothing is nil, save the value
+                                     (not (or (nil? k)
+                                              (nil? v)
+                                              (nil? t)))
+                                     (assoc-in old [t k] v)
+
+                                     ;; value (only) is nil, remove the value (tombstone)
+                                     (and (nil? v)
+                                          (not (or (nil? t)
+                                                   (nil? k))))
+                                     (update old t (fnil #(dissoc % k) {}))
+
+                                     ;; default, return old value
+                                     :else
+                                     (do
+                                       (warn (format "merge-update-with-ktable does not have logic for this case: topic='%s', key='%s', value='%s'"
+                                                     (str t) (str k) (str v)))
+                                       old)))
+                                 (fn [old-meta]
+                                   (cond-> old-meta
+
+                                     ;; If we have an offset and a partition (int) value we will store it,
+                                     ;; keeping the max of all offsets per partition we've encountered.
+                                     ;; `max` throws with `nil` so we need a non-nil value to compare
+                                     ;; the encountered offset with. Choose `-1` because it is less than
+                                     ;; any offset we will encounter (ie invalid offset value) but still
+                                     ;; an int that will not throw inside max.
+                                     (and o p)
+                                     (update-in [:ktable/topic-partition-offsets t p]
+                                                (fnil #(max % o)
+                                                      -1))
+
+                                     ;; in retrospect, `:ktable/recerd-headers` was not a great idea, as it's too limiting
+                                     ;; `:ktable/record-data` is better since it captures all of the kafka record data,
+                                     ;; instead of just some of it.
+                                     (or o p ts hdrs)
+                                     (assoc-in [:ktable/record-data t k] (->> {:offset    o
+                                                                               :partition p
+                                                                               :timestamp ts
+                                                                               :headers   hdrs}
+                                                                              (remove (comp nil? second))
+                                                                              (into {})))
```

Comment:
`:ktable/record-data` is new meta-data on ktables. It's intended to completely replace `:ktable/record-headers`, but that is still in use.